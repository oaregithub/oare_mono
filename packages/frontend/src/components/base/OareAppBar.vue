<template>
  <v-app-bar app dark clipped-left color="#002E5D">
    <v-app-bar-nav-icon @click="$emit('nav-icon-click')" />
    <div id="logo" v-if="$vuetify.breakpoint.smAndUp">
      <a href="https://byu.edu" target="_blank">
        <v-img
          src="/BYU_abbrev.png"
          class="mt-5 mx-3"
          max-height="50px"
          max-width="75px"
        />
      </a>
    </div>

    <div id="leftSide">
      <div id="titleBox">
        <div class="subtitle-1">
          <a
            class="blue-grey--text text--lighten-3 no_underline"
            href="https://history.byu.edu"
            target="_blank"
            >{{ i18n.t('appBar.historyDep') }}</a
          >
        </div>
        <div class="headline">
          <router-link class="white--text no_underline" to="/">{{
            title
          }}</router-link>
        </div>
      </div>
    </div>

    <v-spacer />
    <div>
      <div class="d-flex align-center">
        <v-badge
          v-if="isAdmin"
          :value="displayAdminBadge"
          color="error"
          overlap
          dot
          left
          class="test-admin-badge"
        >
          <v-btn
            v-if="isAdmin"
            class="mr-2 test-admin-btn"
            text
            to="/admin/groups"
            >Admin</v-btn
          >
        </v-badge>
        <v-btn
          v-if="!isAuthenticated"
          class="test-login-btn"
          text
          to="/login"
          >{{ i18n.t('appBar.login') }}</v-btn
        >
        <v-menu v-else offset-y>
          <template v-slot:activator="{ on }">
            <v-btn text v-on="on">
              {{ i18n.t('appBar.welcome') }},
              {{ firstName }}
            </v-btn>
          </template>
          <v-list>
            <v-list-item @click="$router.push('/dashboard/profile')">
              <v-list-item-title>Dashboard</v-list-item-title>
            </v-list-item>
            <v-list-item @click="logout">
              <v-list-item-title>{{
                i18n.t('appBar.logout')
              }}</v-list-item-title>
            </v-list-item>
          </v-list>
        </v-menu>
      </div>
    </div>

    <template #extension>
      <v-row class="d-flex justify-center">
        <v-menu
          offset-y
          open-on-hover
          v-if="
            permissions.includes('WORDS') ||
            permissions.includes('NAMES') ||
            permissions.includes('PLACES')
          "
        >
          <template v-slot:activator="{ on, attrs }">
            <v-btn color="primary" dark v-bind="attrs" v-on="on">
              Lexica
            </v-btn>
          </template>
          <v-list>
            <v-btn
              class="test-words"
              text
              to="/words/A"
              v-if="permissions.includes('WORDS')"
              >Words</v-btn
            >
            <br />
            <v-btn
              class="test-names"
              text
              to="/names/A"
              v-if="permissions.includes('NAMES')"
              >Names</v-btn
            >
            <br />
            <v-btn
              class="test-places"
              text
              to="/places/A"
              v-if="permissions.includes('PLACES')"
              >Places</v-btn
            >
          </v-list>
        </v-menu>
        <v-btn
          class="test-places"
          text
          to="/people/A"
          v-if="permissions.includes('PEOPLE')"
          >People</v-btn
        >
        <v-btn class="test-texts" text to="/collections/A-J">Texts</v-btn>
        <v-btn class="test-search" text to="/search/texts">Search</v-btn>

        <v-menu offset-y open-on-hover>
          <template v-slot:activator="{ on, attrs }">
            <v-btn color="primary" dark v-bind="attrs" v-on="on"> Misc. </v-btn>
          </template>
          <v-list>
            <v-btn text to="/about" width="100%">About</v-btn>
            <br />
            <v-btn text to="/tutorial" width="100%">Tutorial</v-btn>
          </v-list>
        </v-menu>
      </v-row>
    </template>
  </v-app-bar>
</template>

<script lang="ts">
import { defineComponent, computed, PropType } from '@vue/composition-api';
import VueI18n from 'vue-i18n';
import Router from 'vue-router';
import defaultI18n from '../../i18n/index';
import sl from '@/serviceLocator';

export default defineComponent({
  name: 'OareAppBar',
  props: {
    router: {
      type: Object as PropType<Router>,
      required: true,
    },
    i18n: {
      type: Object as PropType<VueI18n>,
      default: () => defaultI18n,
    },
  },
  setup({ router, i18n }, context) {
    const store = sl.get('store');
    const serverProxy = sl.get('serverProxy');

    const title = computed(() => {
      if (context.root.$vuetify.breakpoint.smAndDown) {
        return 'OARE';
      }
      return i18n.t('appBar.oare');
    });

    const displayAdminBadge = computed(() =>
      Object.values(store.getters.displayAdminBadge).includes(true)
    );

    const isAdmin = computed(() => store.getters.isAdmin);
    const isAuthenticated = computed(() => store.getters.isAuthenticated);
    const firstName = computed(() =>
      store.getters.user ? store.getters.user.firstName : ''
    );
    const permissions = computed(() =>
      store.getters.permissions.map(permission => permission.name)
    );

    const logout = () => {
      store.logout();
      router.push('/login');
      serverProxy.logout();
    };

    return {
      title,
      logout,
      isAdmin,
      isAuthenticated,
      firstName,
      permissions,
      displayAdminBadge,
    };
  },
});
</script>

<style scoped>
.no_underline {
  text-decoration: none;
}

#appBarContainer {
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: space-around;
  background-color: #002e5d;
  height: 0.66in;
  align-items: center;
}
#leftSide {
  display: flex;
  flex-direction: row;
  align-items: flex-start;
  justify-content: left;
}
#titleBox {
  display: flex;
  flex-direction: column;
  align-content: center;
  margin-left: 10px;
}
#logo {
  height: 0.66in;
}
</style>
