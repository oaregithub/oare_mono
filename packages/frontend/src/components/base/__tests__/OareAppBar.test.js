import Vuetify from "vuetify";
import VueCompositionApi from "@vue/composition-api";
import { mount, createLocalVue } from "@vue/test-utils";
import OareAppBar from "@/components/base/OareAppBar";

const vuetify = new Vuetify();
const localVue = createLocalVue();
localVue.use(VueCompositionApi);

describe("OareAppBar.vue", () => {
  let user = {
    first_name: "Test",
    last_name: "User",
    is_admin: false,
    email: "test@email.com",
    id: 0,
  };

  let getters = {
    user,
    isAuthenticated: true,
  };

  const mockProps = {
    store: {
      dispatch: jest.fn(),
    },
    router: {
      push: jest.fn(),
    },
    i18n: {
      t: jest.fn(),
      locale: "us",
    },
  };

  const createWrapper = () =>
    mount(OareAppBar, {
      localVue,
      vuetify,
      mocks: {
        $store: {
          getters,
        },
        // $i18n: {},
      },
      propsData: mockProps,
      stubs: ["router-link"],
    });

  it("matches snapshot", () => {
    expect(createWrapper()).toMatchSnapshot();
  });

  it("doesn't show Admin button when user is not admin", () => {
    user.is_admin = false;
    expect(createWrapper().find("[data-admin-btn]").exists()).toBe(false);
  });

  it("shows Admin button when user is admin", () => {
    user.is_admin = true;
    expect(createWrapper().find("[data-admin-btn]").exists()).toBe(true);
  });

  it("shows Login button when not logged in", () => {
    getters.isAuthenticated = false;
    expect(createWrapper().find("[data-login-btn]").exists()).toBe(true);
  });
});
