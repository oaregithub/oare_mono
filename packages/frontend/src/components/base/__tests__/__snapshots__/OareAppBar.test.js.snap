// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OareAppBar.vue matches snapshot 1`] = `
<header class="v-sheet theme--dark v-toolbar v-toolbar--extended v-app-bar v-app-bar--clipped v-app-bar--fixed" style="height: 112px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px; background-color: rgb(0, 46, 93); border-color: #002e5d;">
  <div class="v-toolbar__content" style="height: 64px;"><button type="button" class="v-app-bar__nav-icon v-btn v-btn--flat v-btn--icon v-btn--round theme--dark v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu theme--dark"></i></span></button>
    <div id="logo"><a href="https://byu.edu" target="_blank">
        <div class="v-image v-responsive mt-5 mx-3 theme--dark" style="max-height: 50px; max-width: 75px;">
          <transition-stub name="fade-transition" mode="in-out">
            <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;"></div>
          </transition-stub>
          <div class="v-responsive__content"></div>
        </div>
      </a></div>
    <div id="leftSide">
      <div id="titleBox">
        <div class="subtitle-1"><a href="https://history.byu.edu" target="_blank" class="blue-grey--text text--lighten-3 no_underline"></a></div>
        <div class="headline">
          <router-link-stub to="/" class="white--text no_underline"></router-link-stub>
        </div>
      </div>
    </div>
    <div class="spacer"></div>
    <div>
      <div class="v-menu"><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--small"><span class="v-btn__content"><span class="flag-icon flag-icon-us"></span></span></button>
        <!---->
      </div>
      <!---->
      <div class="v-menu"><button type="button" class="v-btn v-btn--flat v-btn--text theme--dark v-size--default"><span class="v-btn__content">
          ,
          Test
        </span></button>
        <!---->
      </div>
    </div>
  </div>
  <div class="v-toolbar__extension" style="height: 48px;">
    <div class="row d-flex justify-center">
      <!---->
      <!---->
      <!---->
      <router-link-stub class="v-btn v-btn--flat v-btn--router v-btn--text theme--dark v-size--default"><span class="v-btn__content">Texts</span></router-link-stub>
      <router-link-stub class="v-btn v-btn--flat v-btn--router v-btn--text theme--dark v-size--default"><span class="v-btn__content">Search</span></router-link-stub>
    </div>
  </div>
</header>
`;
